### PR TITLE
DAOS-10660 control: Remove duplicate rank results after system fanout (#9204)

### PR DESCRIPTION
**Pull request by tanabarr**
There are situations where multiple results for the same rank maybe
received after a distributed system fanout request e.g. during
dmg system stop command execution.

Work-around this issue by removing duplicate results for a single rank
after the fanout results have been collected. Log an error if the
duplicate results do not have identical content.

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: src/control/server/mgmt_system.go [hunk 1]
```diff
@@ -494,6 +495,49 @@ func (svc *mgmtSvc) resolveRanks(hosts, ranks string) (hitRS, missRS *system.Ran
...
+
+	if len(seenResults) == len(resp.Results) {
+		return
+	}
+
+	newResults := make(system.MemberResults, 0, len(seenResults))
+	for _, res := range seenResults {
+		newResults = append(newResults, res)
+	}
+	resp.Results = newResults
```
Comment (kjacque) [inline src/control/server/mgmt_system.go:538]: Any reason not to set `resp.Results = seenResults` and avoid the extra work?

**Review by kjacque — COMMENTED** [review]
I guess I'm struggling a bit to understand this specific case. Is the issue that we're creating "unresponsive" results for ranks that actually returned a result? Or is there another way we can get dupes?

**Comment by tanabarr** [comment]
> I guess I'm struggling a bit to understand this specific case. Is the issue that we're creating "unresponsive" results for ranks that actually returned a result? Or is there another way we can get dupes?

No the duplication is not related to generation of "unresponsive" Rank results (which occurs in the case that a harness is inaccessible and so synthetic results need to be generated for each of its resident ranks). I have explained in some detail the situation that causes the duplicate Rank results in this JIRA comment: https://daosio.atlassian.net/browse/DAOS-10660?focusedCommentId=98455 and here is the summary given in the ticket: "In summary it turns out it's a subtle (unusual) configuration difference that exposes a bug where, because the ranks are requesting join over respective fabric interfaces, 2 ranks on the same host end up as entries in the system membership with different management IP addresses (because the management addresses are actually on the fabric network), and the result is duplicate stop results in the response for the ranks on the MS host. A more detailed analysis and recommendations are provided in the comments.". This situation is distinct from the normal case where both ranks that share the same host join from the same IP address and the fanout is then only executed once on the given host (the fanout executes a system stop on each of the host addresses gathered from deduped list of rank join request origin addresses).

I hope that explains the situation.

File: src/control/server/mgmt_system.go (hunk 1 above)
Comment (tanabarr) [reply src/control/server/mgmt_system.go:538]: seenResults is a map and resp.Results is a slice

**Comment by kjacque** [comment]
@tanabarr I see, thanks for the explanation. It seems like this is basically the same issue as [DAOS-10661](https://daosio.atlassian.net/browse/DAOS-10661), where we are sending the Join request from any interface on the system that can find a route to the destination address, and not being picky about the source. I do wonder if this would be resolved by canonizing the management address (which is what I'm working on now).

**Review by kjacque — APPROVED** [review]

**Comment by tanabarr** [comment]
> @tanabarr I see, thanks for the explanation. It seems like this is basically the same issue as [DAOS-10661](https://daosio.atlassian.net/browse/DAOS-10661), where we are sending the Join request from any interface on the system that can find a route to the destination address, and not being picky about the source. I do wonder if this would be resolved by canonizing the management address (which is what I'm working on now).

Yes, as noted in the ticket the root cause is different interfaces being used in join from the same host. I suggested in the ticket that the root cause, and existence of duplicate rank results, should both be fixed.

**Review by kjacque — APPROVED** [review]

**Pull request merged**